Move DataTypes.DataDrivenTests to .Net Core

hansmbakker opened the pull request:
Modernize the `Microsoft.Recognizers.Text.DataTypes.DataDrivenTests` and `Microsoft.Recognizers.Text.DataDrivenTests`.

tellarin commented:
Same compat CI error.

tellarin commented:
Thanks for tracking the ```DataSourceAttribute``` dependency in .net core.

hansmbakker commented:
What is the route you would suggest to take?
- use his workaround https://github.com/Microsoft/testfx/issues/233#issuecomment-432644083
- internally contact @cltshivash (see https://github.com/Microsoft/testfx/issues/233#issuecomment-445106443)
- wait for MS to unblock this by implementing ` DataSourceAttribute` in .net core.
- something else?

hansmbakker commented:
> To avoid VS getting confused during test runs.

I had no problems when running the tests, did you encounter issues?

hansmbakker commented:
I rebased the branch on master and made the datadriventests project readable using the new csproj syntax

hansmbakker commented on .NET/Microsoft.Recognizers.Text.DataDrivenTests/TestHelpers.cs:
```diff
@@ -33,7 +33,7 @@ public static void InitFromTestContext(this TestResources resources, TestContext
             var recognizerLanguage = className.Split('_');
 
-            var directorySpecs = Path.Combine("..", "..", "..", "..", "Specs", recognizerLanguage[0], recognizerLanguage[1]);
+            var directorySpecs = Path.Combine("..", "..", "..", "..", "..", "Specs", recognizerLanguage[0], recognizerLanguage[1]);
```
I'm not sure whether this extra level up in the directory tree was needed because of the csproj format update or because Visual Studio 2019 handles working paths differently.

hansmbakker commented:
Could you please explain which script breaks, and what the dependencies of that script are? Is there an install script to setup those dependencies?

I assume you mean `CreateAllPackages.cmd` in the .NET folder?

hansmbakker commented:
It works for me when I do the following steps:
- Open `Developer Command Prompt for VS2019`
- `cd` into the `Recognizers-Text\.NET` folder
- run `build.ci.cmd`
- run `CreateAllPackages.cmd`

The result is:
```
..
*** Building Microsoft.Recognizers.Text
Het systeem kan het opgegeven pad niet vinden.
Het systeem kan het opgegeven pad niet vinden.
Attempting to build package from 'Microsoft.Recognizers.Text.nuspec'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.1.0.0.nupkg'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.1.0.0.symbols.nupkg'.
*** Succeeded to build Microsoft.Recognizers.Text
*** Building Microsoft.Recognizers.Text.Number
Het systeem kan het opgegeven pad niet vinden.
Het systeem kan het opgegeven pad niet vinden.
Attempting to build package from 'Microsoft.Recognizers.Text.Number.nuspec'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.Number.1.0.0.nupkg'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.Number.1.0.0.symbols.nupkg'.
*** Succeeded to build Microsoft.Recognizers.Text.Number
*** Building Microsoft.Recognizers.Text.NumberWithUnit
Het systeem kan het opgegeven pad niet vinden.
Het systeem kan het opgegeven pad niet vinden.
Attempting to build package from 'Microsoft.Recognizers.Text.NumberWithUnit.nuspec'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.NumberWithUnit.1.0.0.nupkg'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.NumberWithUnit.1.0.0.symbols.nupkg'.
*** Succeeded to build Microsoft.Recognizers.Text.NumberWithUnit
*** Building Microsoft.Recognizers.Text.DateTime
Het systeem kan het opgegeven pad niet vinden.
Het systeem kan het opgegeven pad niet vinden.
Attempting to build package from 'Microsoft.Recognizers.Text.DateTime.nuspec'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.DateTime.1.0.0.nupkg'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.DateTime.1.0.0.symbols.nupkg'.
*** Succeeded to build Microsoft.Recognizers.Text.DateTime
*** Building Microsoft.Recognizers.Text.Sequence
Het systeem kan het opgegeven pad niet vinden.
Het systeem kan het opgegeven pad niet vinden.
Attempting to build package from 'Microsoft.Recognizers.Text.Sequence.nuspec'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.Sequence.1.0.0.nupkg'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.Sequence.1.0.0.symbols.nupkg'.
*** Succeeded to build Microsoft.Recognizers.Text.Sequence
*** Building Microsoft.Recognizers.Text.Choice
Het systeem kan het opgegeven pad niet vinden.
Het systeem kan het opgegeven pad niet vinden.
Attempting to build package from 'Microsoft.Recognizers.Text.Choice.nuspec'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.Choice.1.0.0.nupkg'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.Choice.1.0.0.symbols.nupkg'.
*** Succeeded to build Microsoft.Recognizers.Text.Choice
*** Building Microsoft.Recognizers.Text.DataTypes.TimexExpression
Het systeem kan het opgegeven pad niet vinden.
Het systeem kan het opgegeven pad niet vinden.
Attempting to build package from 'Microsoft.Recognizers.Text.DataTypes.TimexExpression.nuspec'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.DataTypes.TimexExpression.1.0.0.nupkg'.
Successfully created package '..\nuget\Microsoft.Recognizers.Text.DataTypes.TimexExpression.1.0.0.symbols.nupkg'.
*** Succeeded to build Microsoft.Recognizers.Text.DataTypes.TimexExpression
```

This is the same result as what I see in the build logs for a [successful PR](https://github.com/Microsoft/Recognizers-Text/pull/1053) - see https://ci.appveyor.com/project/tellarin/recognizers-text/builds/20891640?fullLog=true#L18482

How do you suggest to continue? 
- I removed the `vswhere` and `NuGet.Console` packages from the `Microsoft.Recognizers.Text.DataDrivenTests` because they I believe they are not being packaged. I assume that doesn't have influence.
- I did not change the output folders of the projects other than the test projects.
- I find it difficult to see where further to investigate on the build server, other than modifying the `CreatePackage.cmd` scripts so that they will log more verbose.

hansmbakker commented:
@iMicknl @tellarin I found a way to port Microsoft.Recognizers.Text.DataDrivenTests.csproj to .Net Core.

Based on what I see in https://github.com/Microsoft/testfx/issues/233, we unfortunately cannot expect `DataSourceAttribute` to be available in .Net Core any time soon. I found `DynamicDataAttribute` too limiting* and therefore I created my own attribute that I called `NetCoreTestDataSourceAttribute` for now.

I removed the writing to csv since that is not needed anymore - the TestModel objects are directly passed to the test methods.

One refactoring point is that I'm not particularly fond of the `TestSpec` property - it could simply be a parameter to the actual test implementation method, rather than a class property. However that refactoring can probably be done separately of this .Net Core PR.

I see that AppVeyor still doesn't pass (for another reason than before), even though the tests run... I unfortunately don't see what's wrong and would like to get your feedback.

*reason that `DynamicDataAttribute` is too limiting is that you cannot implement a generic data loading method that loads data based on the test method or based on parameters with it. The test-specific `TestContext` property is only set after `DynamicDataAttribute`s `GetData(…)` method is called. Also the test-specific `TestContext` is non-static which means that you would need to forward that to another static property (because `DynamicDataAttribute` expects a static method or property).

hansmbakker commented:
@iMicknl thanks! No, I don't have a clue unfortunately.

Besides the CIs failing, there might be some optimization possibilities:
- Azure DevOps offers free CI pipelines for public github repos, would this allow using one CI that offers insights to non-maintainers? (maybe you have a good reason for using two different CIs but it looks like double maintenance for me?)
- maybe a dedicated .Net Core test runner template could be used?
- the build process might be possible without the `robocopy` script? I see a lot of files being copied around?

hansmbakker commented:
@tellarin could you please have a look at this so that it does not get outdated?

tellarin commented:
Hi @hansmbakker,
I should have time to take a closer look at it and re-run tests in the next couple days. Sorry for the delay. 
Would you have time to look into the conflicts? They should be minor, I think.

tellarin commented:
@JuanAr could you folks help resolve the conflicts here? It seems most come from the recent StyleCop changes.

hansmbakker commented:
@tellarin: I’m sorry, at least the coming week unfortunately I won’t have spare time to fix merge conflicts (because I cannot spend time on this at work).

However if anybody wants to help out (I checked “allow edits from maintainers”) then I would very much appreciate it; besides I think it would help this project to move forward (since then mac and linux users could participate easier).

Otherwise I can fix it when I have more free time again.

hansmbakker commented:
@tellarin @JuanAr I made time and fixed the merge conflicts.

I believe there is a high chance of merge conflicts appearing again because the test class structure changed and the signature of the test methods changed (due to how the test spec is injected in the test method in .Net Core). Therefore I would really appreciate it if somebody could look at this PR before new merge conflicts appear.

hansmbakker commented:
@iMicknl would you know whether there is any other maintainer who can review this?

This PR is waiting for a review for >1 month now (since https://github.com/Microsoft/Recognizers-Text/pull/1050#issuecomment-447609178) and merge conflicts appeared for the second time because other PRs were merged while this one is kept open...

I understand people are probably busy on other things but unfortunately this makes me reluctant to spend more of my private time on it unless I know that the time I spend on it is worth it.

tellarin commented:
Hi @hansmbakker,
we really appreciate the contribution. I didn't merge it as the PR build failed last time.
I can help solve the conflicts soon.

tellarin commented:
> * the build process might be possible without the `robocopy` script? I see a lot of files being copied around?

This is something we plan to remove. The idea is to have the projects with no copy and build.cmd/build.sh will copy things to the right place.

tellarin commented:
> * Azure DevOps offers free CI pipelines for public github repos, would this allow using one CI that offers insights to non-maintainers? (maybe you have a good reason for using two different CIs but it looks like double maintenance for me?)

Unfortunately there are some non-public tests to be ran, so we can't yet make that part public. We'll look into alternatives to avoid this as sometimes build failures can block people.

tellarin commented:
@hansenms it seems there's something strange with the new way you load data for the tests.

In 7d9519d, the AppVeyor build fails in 4 tests. But the VSO CI fails with: 
"2019-01-18T22:42:58.3912389Z Total tests: 14190. Passed: 11524. Failed: 8. Skipped: 2658."
In the current commit, AppVeyor fails 11 tests and VSO fails with: 
"2019-01-22T10:14:09.9910775Z Total tests: 14204. Passed: 11543. Failed: 4. Skipped: 2657."

Perhaps there's something that doesn't guarantee execution order going on?

hansmbakker commented:
@tellarin thank you for helping here! I understand it was not merged yet until we understood the build issues, the issue with this PR was just that I could not take it further on my own.

> it seems there's something strange with the new way you load data for the tests

The only thing I could think of is that I removed the output directories (see https://github.com/Microsoft/Recognizers-Text/pull/1050#discussion_r240350312) in the `.csproj`, and I added `..` to the test spec file paths.

> Perhaps there's something that doesn't guarantee execution order going on?

I have no idea about that, unfortunately. The most obvious difference between local and automated build I could see were the `robocopy` steps. The difference between commits and especially between build servers is a black box for me. I cannot look further into the build output I'm afraid - appveyor is the only one that lets me load the page, and it only shows the build log but no clear overview of tests (the tests page in appveyor is empty for me). Also I cannot see what made the tests fail.

tellarin commented:
Close/reopen to rebuild and check the logs after the changes in CI build scripts.

tellarin commented:
> > Perhaps there's something that doesn't guarantee execution order going on?
> 
> I have no idea about that, unfortunately. The most obvious difference between local and automated build I could see were the `robocopy` steps. The difference between commits and especially between build servers is a black box for me. I cannot look further into the build output I'm afraid - appveyor is the only one that lets me load the page, and it only shows the build log but no clear overview of tests (the tests page in appveyor is empty for me). Also I cannot see what made the tests fail.

It seems the tests fail because the data loaded for a test doesn't match the test being run. Maybe some async issue in the new loader. Does it always work for you locally? For me I also got a few errors locally, not always the same.

Let's check the latest build log as soon as the build finishes.

tellarin commented:
@hansmbakker we had a weird build failure on VSO before, but now it's back.
This PR fails in some new Dutch tests. I'll try to take a look at them tomorrow.
When you look at the build results, the Tests tab on AppVeyor is always empty, but if you check the full logs you should be able to see which specific tests fail in C#.

iMicknl commented:
I also had a look at this today and have some Currency test failing locally (English + Spanish), will try to give it a look again with the changes @tellarin made in the Dutch tests. (VS2017 CE)

iMicknl commented:
@tellarin @hansmbakker I fixed all conflicts and fixed some StyleCop issues, the others I ignored for now. Still facing the same issues with CurrencyModel, can't get a hold of it. Two screenshots with errors, without changing other code than the specs. Only CurrencyModel in different languages is failing. (and German DateTimeModelSplitDateAndTime, but I think that one is not implemented).

<img width="596" alt="Screenshot 2019-03-27 at 19 42 32" src="https://user-images.githubusercontent.com/1424596/55110460-5c1be200-50d8-11e9-9484-5d7e63851b15.png">
<img width="420" alt="Screenshot 2019-03-27 at 21 09 27" src="https://user-images.githubusercontent.com/1424596/55110461-5de5a580-50d8-11e9-87b4-430f914abb0f.png">

hansmbakker commented:
@iMicknl thank you for pushing this forward!

With the screenshot from `TestNumberWithUnit_English` --> `CurrencyModel` you shared I see that `sen` is defined twice in `English - NumbersWithUnit.yaml`: once for `Euro - Cent` and once for `Sen`. `Sen` is also defined in the `CurrencyPreffixList` in `English` --> `NumbersWithUnitDefinitions.cs`

For me the `actualResults` contains "unit = Sen; value = 100000" (looks correct to me seeing an input string of `sen 100,000`).
For me the `expectedResults` contains "unit = Cent; value = 100000" (which matches the test spec json) but frankly speaking that looks like a test spec error rather than an implementation error to me?

I do not know enough about Recognizers-Text to quickly see whether it should prefer `Sen` or `Euro - Cent` because of the `en-us` culture?

I do see other broken tests, but I thought these tests were just broken because of the code they test and maybe the culture issue in #774.

I debugged `TestNumber_English` --> `NumberModel` and I added my findings to that existing bug. If there are more of such issues we'll see different tests breaking.

iMicknl commented:
@hansmbakker it looks also like a test spec error to me, however this is (for now) the behaviour in this package. Something is going wrong with the Currency tests (in different locales) after this PR and I can't get a hold of it... It is not only the above testcase, but it are multiple locales. (in multiple test runs). My issue is that I can't get a consistent behaviour, it can be different every run.

tellarin commented:
@hansmbakker @iMicknl Yep, the unexpected Currency behaviour only happens in this PR and it was the main reason to block the merge back near the PR creation. I also didn't have time to look into why the Currency issue happens. :(
The number format issue is a known issue and a solution is coming soon.

STayinloves commented:
@hansmbakker @tellarin @iMicknl I have updated this PR to pull in upstream changes. 

The currency issue is caused by unreliable order of `Dictionary` in C# which is widely used in `NumberWithUnit` Recognizer, also by ambiguous attributions in some patterns.

For example in `Spanish-NumbersWithUnit.yaml`, centavo belongs to both `Centavo` and `Centavo argentino`:
```yaml
CurrencySuffixList: !dictionary
  types: [ string, string ]
  entries:
    Centavo: centavo|centavos
    Centavo argentino: centavo argentino|centavos argentinos|centavo|ctvo.|ctvos.
```

This problem can be solved in two way:
- fundamentally switch to using `SortedDictionary` in `NumberWithUnit` Recognizer
- fix ambiguous attributions

tellarin commented:
Thanks for tracking this, @STayinloves!
We should definitely fix the ambiguous attributions. In the "centavo" case, it should not be by itself in the Argentinian entry.
Regarding ```SortedDictionary```, the cost of the change would happen only at construction, correct? If so, I think it's safe to make the change.

STayinloves commented:
Hi @tellarin, I have updated the PR to merge in upstream changes, the appveyor CI passed, but I have no clue about the failure of Azure CI. Could you help to trace that? Thanks!

tellarin commented:
It fails with one German test. The log says:

> Failed   Microsoft.Recognizers.Text.DateTime.Tests.TestDateTime_German HolidayParser
Error Message:
 Assert.IsNotNull failed. Input: "Ich freue mich schon auf Weihnachten."
Stack Trace:
   at Microsoft.Recognizers.Text.DataDrivenTests.TestBase.TestDateTimeParser(TestModel testSpec) in D:\a\1\s\.NET\Microsoft.Recognizers.Text.DataDrivenTests\TestBase.cs:line 195

But no more details. :-/

tellarin commented:
Re-triggering builds.

tellarin commented on Patterns/German/German-DateTime.yaml:
```diff
@@ -966,7 +966,6 @@ HolidayNames: !dictionary
     girlsday: [ girlsday ]
     whiteloverday: [ whiteloverday ]
     loverday: [ loverday ]
-    christmas: [ weihnachten, weihnachtstag ]
```
Why this?

STayinloves replied:
There’s already a `firstchristmasday` entry containing the same values.

tellarin commented:
@STayinloves Now it's even more strange, no specific test failures, but:
> Total tests: Unknown. Passed: 14348. Failed: 0. Skipped: 2821.
Test Run Aborted.
Test execution time: 2.0782 Minutes
.NET Test failure(s) found

tellarin commented:
The failure seemed to be caused by behaviour specific to dotnet core where test discovery failed with semi-silent issues:
- "Assets file project.assets.json not found. Run a NuGet package restore" or "Make sure test project has a nuget reference of package "Microsoft.NET.Test.Sdk"" - solved by adding a restore in the call to _msbuild_;
- "Total tests: Unknown." output and test run failure - solved by making sure tests are not loaded from _obj_ directories.

hansmbakker commented:
Great that you're picking up this PR, thank you!

tellarin commented:
@hansmbakker @STayinloves @iMicknl @luizcavalcanti Thank you all for the great effort to move all to multi-plat .NET core. We've finally tracked all pending issues and this PR will be merged very soon.

iMicknl commented:
Great debugging work @tellarin & @STayinloves! :-) 
Thanks @hansmbakker! At least I should be able now to run the tests on my Mac.

hansmbakker commented:
Happy to see this is merged :) hopefully this will enable more people to use and develop for this library!

I really appreciate everybody who committed to this PR to fix the blocking issues since they really got me stuck! Thank you!
